src/data_source/notifier.rs

```diff
@@ -0,0 +1,363 @@
+//!
+//! This is important because subscriptions are requested in response to read-only client requests,
+//! which are supposed to run in parallel as much as possible. By contrast, notifications are
+//! ususally send from internal server tasks (e.g. the background task that updates the data source
+//! when new blocks are committed). It is less of a problem if these internal tasks contend with
+//! each other, because they are not directly blocking responses to clients, and we have more
+//! control over how and when they acquire shared resources.
+//!
+//! This requirement also empowers us to create a simpler design for the high-level fetching logic.
+//! Specifically, we can reuse the same code for fetching individual resources as we use for
+//! long-lived subscription streams (e.g. `subscribe_blocks` is a thin wrapper around
+//! `get_block_range`). We do not have to worry about adding complex logic to reuse notification
+//! subscriptions for long-lived streams, because subscribing anew for each entry in the stream has
+//! low overhead in terms of contention over shared resources -- the dominant caused in any
+//! concurrent channel, after data copying (see above).
+//!
+//! This further lets us simplify the interface of this channel a bit: since all notifications are
+//! oneshot, consumers deal with futures rather than streams.
+//!
+//! This requirement is satisfied by maintaining the list of subscribers to a [`Notifier`] in a way
+//! that moves most resource contention to message senders, rather than receivers. We make the
+//! assumption that there is less concurrency among senders. In the common case, there is just one
+//! sender: the task monitoring HotShot for new blocks. Occasionally, there may be other tasks
+//! spawned to fetch missing resources and send them through the [`Notifier`], but these should be
+//! relatively few and rare.
+
+use async_compatibility_layer::channel::{
+    oneshot, unbounded, OneShotReceiver, OneShotSender, UnboundedReceiver, UnboundedSender,
+};
+use async_std::sync::Arc;
+use derivative::Derivative;
+use futures::future::{BoxFuture, FutureExt};
+use std::{
+    future::IntoFuture,
+    sync::atomic::{AtomicBool, Ordering},
+};
+
+pub trait Predicate<T>: 'static + Send + Sync + Fn(&T) -> bool {}
+impl<F, T> Predicate<T> for F where F: 'static + Send + Sync + Fn(&T) -> bool {}
+
+#[derive(Derivative)]
+#[derivative(Debug)]
+struct Subscriber<T> {
+    #[derivative(Debug = "ignore")]
+    predicate: Box<dyn Predicate<T>>,
+    #[derivative(Debug = "ignore")]
+    sender: Option<OneShotSender<T>>,
+    closed: Arc<AtomicBool>,
+}
+
+impl<T> Subscriber<T> {
+    fn is_closed(&self) -> bool {
+        // A subscriber can be closed because it was explicitly closed by its receiver (e.g. the
+        // receiver was dropped)
+        self.closed.load(Ordering::Relaxed) ||
+        // Or because it has already been notified, which `take`s the oneshot sender.
+        self.sender.is_none()
```

Comment:
Small nit but maybe a bit more efficient to put the is_none() check first?

Reply:
Fair enough